lib/evaluate/host.h

```diff
@@ -0,0 +1,229 @@
+          std::numeric_limits<float>::is_iec559,
+      float, UnsupportedType>;
+};
+
+template<> struct HostTypeHelper<Type<TypeCategory::Real, 8>> {
+  // IEE 754 64bits
+  using Type = std::conditional_t<sizeof(double) == 8 &&
+          std::numeric_limits<double>::is_iec559,
+      double, UnsupportedType>;
+};
+
+template<> struct HostTypeHelper<Type<TypeCategory::Real, 10>> {
+  // X87 80bits
+  using Type = std::conditional_t<sizeof(long double) >= 10 &&
+          std::numeric_limits<long double>::digits == 64 &&
+          std::numeric_limits<long double>::max_exponent == 16384,
+      long double, UnsupportedType>;
+};
+
+template<> struct HostTypeHelper<Type<TypeCategory::Real, 16>> {
+  // IEE 754 128bits
+  using Type = std::conditional_t<sizeof(long double) == 16 &&
+          std::numeric_limits<long double>::digits == 113 &&
+          std::numeric_limits<long double>::max_exponent == 16384,
+      long double, UnsupportedType>;
+};
+
+template<int KIND> struct HostTypeHelper<Type<TypeCategory::Complex, KIND>> {
+  using RealT = Fortran::evaluate::Type<TypeCategory::Real, KIND>;
+  using Type = std::conditional_t<HostTypeExists<RealT>(),
+      std::complex<HostType<RealT>>, UnsupportedType>;
+};
+
+template<int KIND> struct HostTypeHelper<Type<TypeCategory::Character, KIND>> {
+  using Type = std::conditional_t<KIND <= 8, std::uint8_t, UnsupportedType>;
+};
+
+template<> struct HostTypeHelper<Type<TypeCategory::Character, 1>> {
+  using Type = std::string;
+};
+
+template<> struct HostTypeHelper<Type<TypeCategory::Character, 2>> {
+  using Type = std::u16string;
+};
+
+// Type mapping from host types to F18 types. This need to be placed after all
+// HostTypeHelper specializations.
+template<typename T, typename... TT> struct IndexInTupleHelper {};
+template<typename T, typename... TT>
+struct IndexInTupleHelper<T, std::tuple<TT...>> {
+  static constexpr int value{common::TypeIndex<T, TT...>};
+};
+struct UnknownType {};  // the host type does not match any F18 types
+template<typename HOST_T> struct FortranTypeHelper {
+  using HostTypeMapping =
+      common::MapTemplate<HostType, AllIntrinsicTypes, std::tuple>;
+  static constexpr int index{
+      IndexInTupleHelper<HOST_T, HostTypeMapping>::value};
+  using Type = std::conditional_t<index >= 0,
+      std::tuple_element_t<index, AllIntrinsicTypes>, UnknownType>;
+};
+
+template<typename HOST_T>
+using FortranType = typename FortranTypeHelper<HOST_T>::Type;
+
+template<typename... HT> constexpr inline bool FortranTypeExists() {
+  return (... && (!std::is_same_v<FortranType<HT>, UnknownType>));
+}
+
+// Utility to find "bigger" types that exist on host. By bigger, it is meant
+// that the bigger type can represent all the values of the smaller types
+// without information loss.
+template<TypeCategory cat, int KIND> struct NextBiggerReal {
+  using Type = void;
+};
+template<TypeCategory cat> struct NextBiggerReal<cat, 2> {
+  using Type = Fortran::evaluate::Type<cat, 3>;
```

Comment:
Conversions from IEEE 16-bit real to the truncated 32-bit real will lose precision.

Reply:
Now mapped to IEEE 32bits